### ISSUE TITLE
Support bzlmod-based builds

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#11485]
Bazel is adding a new (better) way to handle dependencies:

https://bazel.build/external/overview#bzlmod

This automatically handles recursive dependencies, resolves diamond dependencies, etc.  And supports multiple registries of packages:

https://bazel.build/external/registry

With a default "global registry":

https://github.com/bazelbuild/bazel-central-registry

We need to figure out a number of things, and implement a number of things.  This list is probably not exhaustive, it needs to be filled out:

- [x] We need to figure out how to have backup sources for our dependencies.  Consider https://blog.aspect.build/configuring-bazels-downloader
- [x] The `curl` module in BCR does not support HTTPs. Filed https://github.com/bazelbuild/bazel-central-registry/issues/2332
- [x] #14454
- [x] We need to decide how to manage the `googleapis` module in BCR.
   - An override for development is fine, but does not work for downstream consumers.  
   - Maybe we can create a new version of `googleapis` in BCR just before creating a new version of `google-cloud-cpp` in BCR.
- [x] We would need at least one CI build using the bzlmod configuration
- [x] Publish initial google-cloud-cpp module to BCR
- [x] Validate google-cloud-cpp module from BCR

**Comment by coryan** [comment]
Waiting for Bazel 7.0

**Comment by alevenberg** [comment]
I'll take a stab at some of the tasks this week.

**Comment by alevenberg** [comment]
The migration guide https://bazel.build/external/migration

**Comment by alevenberg** [comment]
I think we're gonna first just want to compile with bazel 7.0, and then we can tackle bazelmod.

Waiting to see if we can get that build working on Mac (https://github.com/bazelbuild/bazel/issues/20513) and Windows.

**Comment by alevenberg** [comment]
FYI timeline for support for Bazel 6 is until December 2025

**Comment by scotthart** [comment]
Add this to 2024 planning.

**Comment by mering** [comment]
As there is a working MODULE:bazel file already, I would appreciate this being available in BCR. The first uploads could be manually while future ones could be more automated. Do you plan on adding it yourself or should I go ahead an propose a first upload to BCR?

**Comment by scotthart** [comment]
We're still waiting on several of our dependencies to be added to BCR. The Bazel team is aware of this and is working on it.

**Comment by drqec** [comment]
Is it possible at all to use this repository with Bzlmod? I'm trying to depend on it using the following:

```
bazel_dep(name = "google_cloud_cpp", version = GOOGLE_CLOUD_VERSION)
archive_override(
      module_name = "google_cloud_cpp",
      integrity = GOOGLE_CLOUD_INTEGRITY,
      strip_prefix = "google-cloud-cpp-%s" % GOOGLE_CLOUD_VERSION,
      urls = ["https://github.com/googleapis/google-cloud-cpp/archive/v%s.tar.gz" % GOOGLE_CLOUD_VERSION],
)
```

But I get the error:
`Error computing the main repository mapping: module not found in registries: googleapis@0.0.0`

Should projects that want to depend on this continue using a WORKSPACE file?

**Comment by jinseopkim0** [comment]
Created separate issue https://github.com/googleapis/google-cloud-cpp/issues/15296 to track documentation efforts, removed documentation tasks in the issue description.

**Comment by jinseopkim0** [comment]
Added the following tasks to the issue description.
* Publish initial google-cloud-cpp module to BCR
* Validate google-cloud-cpp module from BCR

**Comment by jinseopkim0** [comment]
https://github.com/googleapis/google-cloud-cpp/pull/15327 sent out to validate google-cloud-cpp module from BCR

**Comment by jinseopkim0** [comment]
Tested as consumer of BCR module [3.0.0-rc1 version](https://screenshot.googleplex.com/AxDE2zV3ziNGGTY) using https://github.com/googleapis/google-cloud-cpp/pull/15354 locally.

**Comment by jinseopkim0** [comment]
Tested as consumer of BCR module 3.0.0-rc1 version (https://registry.bazel.build/modules/google_cloud_cpp) using https://github.com/googleapis/google-cloud-cpp/pull/15354 locally.
